pkg/reconciler/autoscaling/versionwrapper/controller.go

```diff
@@ -0,0 +1,68 @@
+package versionwrapper
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"k8s.io/client-go/discovery"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/serving/pkg/reconciler/autoscaling/hpa"
+	hpav2beta2 "knative.dev/serving/pkg/reconciler/autoscaling/hpav2beta2"
+)
+
+func NewController(
+	ctx context.Context,
+	cmw configmap.Watcher,
+) *controller.Impl {
+	kc := kubeclient.Get(ctx)
+	// Starting from 4.10 (1.24.0) we can use the new API version (autoscaling/v2) of HorizontalPodAutoscaler
+	// As we also need to support 4.8 we also need provide the controller using the old API version (autoscaling/v2beta2)
+	if err := CheckMinimumVersion(kc.Discovery(), "1.24.0"); err == nil {
+		return hpa.NewController(ctx, cmw)
+	} else {
+		return hpav2beta2.NewController(ctx, cmw)
+	}
+}
+
+// CheckMinimumVersion checks if current K8s version we are on is higher than the one passed.
+// An error is returned if the version is lower.
+// Based on implementation in SO: https://github.com/openshift-knative/serverless-operator/blob/main/openshift-knative-operator/pkg/common/api.go#L134
+func CheckMinimumVersion(versioner discovery.ServerVersionInterface, version string) error {
```

Comment:
I'm curious if using CheckMinimumVersion from the vendor would also cause cycle? https://github.com/openshift-knative/serving/pull/177/files#diff-d36667190767133e4969de666038f9f5cf65f687f1d449357082dbdfb1c79deaR169
I understand it creates a cycle the other way round but accessing the vendor's function from here should work. Or maybe not? Just checking.

Reply:
That seems to work. Tests are running.